add sign functionality

diveshuttam opened the pull request:
I have changed the commands using basename as otherwise the bash process was seperate and other egpg functions like `call cmd_sign` etc. were not accessible.

There is one thing left in here, for all gui commands we need to change pinentry-tty to something like pinentry-gtk or some other graphical user interface.

diveshuttam committed:
add sign functionality

dashohoxha commented:
> I have changed the commands using basename as otherwise the bash process was seperate and other egpg functions like call cmd_sign etc. were not accessible.

:+1: I did not test it that far.

> There is one thing left in here, for all gui commands we need to change pinentry-tty to something like pinentry-gtk or some other graphical user interface.

On `cmd_init` we set the configuration of gpg-agent to `pinentry-program /usr/bin/pinentry-tty`. Maybe the initialization for the GUI case should be a bit different. If we set no value at all for pinentry, I think that the default value is `pinentry-gtk`.

By the way, we also need to do something about `gpg.conf` because it is deprecated (or useless) in gnupg-2.2.

Another thing is that on the file `config.sh` we need to save a label as well. This is useful for the GUI case but also for the command line. The purpose of this label is to help the user distinguish clearly and easily one key from another (in a scenario when he may need to use more than one key). For example he may use one key for communicating with work colleagues, another one for communicating with friends, and a third one for communicating with family members. Each of these keys lives in a different context (homedir). Normally the user should be able to distinguish them from each-other by the key id (or fingerprint). However this is more difficult and error prone than distinguishing them by a simple and user friendly label.
Maybe we should also allow the user to switch easily the context of egpg from the GUI.

dashohoxha committed:
Option 'set -a' tells Bash to export all variables and functions.

dashohoxha committed:
Ignore output, unless debugging.

diveshuttam commented:
> By the way, we also need to do something about gpg.conf because it is deprecated (or useless) in gnupg-2.2.

I think I have been misinterpreting some of the things here, I again read the discussion on gnupg-devel, It says that keyserver entries are deprecated they should be shifted to dirmngr.conf instead.
So I think other things should work fine.

pinentry-program is to be given in gpg-agent.conf so there should be no problem there too. BTW I am just trying these things.

diveshuttam committed:
add success message and escape error text

error text had <file> etc. in case of usage error in sign
which is considered as pango markup

dashohoxha committed:
Improve messages

dashohoxha commented:
In case that `$file.signed` already exists (it may be a fake or corrupted one) signing fails.
I think that `gpg` in this case tries to ask whether it should override it or not, but since it is not running in interactive mode it fails silently.

dashohoxha commented on src/gui/sign.sh:
```diff
@@ -1,7 +1,9 @@
-    #yad --title="EasyGnuPG | Sign a File" \
-    #   --text="Sign a File"
+    local file output
+    file=$(yad --file --title="EasyGnuPG | Sign a File")
```
We should remove the `$file.signature` in case it exists, or ask the user first whether it should be replaced.

diveshuttam committed:
ask user to overwrite signature

dashohoxha committed:
Improve the logic and messages of gui_sign.

diveshuttam commented on src/gui/sign.sh:
```diff
@@ -1,14 +1,23 @@
-    [[ -f "$file.signature" ]] && message info "Signature saved as $file.signature"
+
+    if [[ -s "$file.signature" ]]; then
+        yad --file --filename="$file.signature" &
```
Hi, I didn't understand the purpose of line 15.
Is it just showing the file to the user? Or you wanted something like user being able to select the desired signature file?

dashohoxha replied:
> Is it just showing the file to the user?

Yes.
